lib/_http_client.js

```diff
@@ -351,7 +362,9 @@ ClientRequest.prototype.abort = function abort() {
   // the request queue through handling in onSocket.
   if (this.socket) {
     // in-progress
```

Comment:
That `this.socket` is defined it doesn't necessarily mean that the request has been sent.

Reply:
I think you should check `req._flushedHeaders` or `req.headerSent` (I don't remember the exact name).

Reply:
Not sure what you mean here, the point here is to just kill the request, regardless whether it has been sent or not.

Reply:
I think you meant `kill the socket`. Wrong. Note that [here (lines 767-779)](https://github.com/nodejs/node/pull/32148/files#diff-e3bc37430eb078ccbafe3aa3b570c91aR767-R779) you free the socket (it goes back to its owner - an instance of Agent) - but if it's already connected you kill it. What's the difference if the request has not been sent? I don't see any.

Reply:
The PR description says:

> Fixed ClientRequest.destroy to dump res and re-use socket in agent
pool aligning it with abort.

As of now this PR doesn't fix this.

```js
const request = https.request('https://example.com');
request.once('socket', () => {
	// The `request.socket` shouldn't be destroyed
	// because the request hasn't been sent.
	// Instead, `request.socket.emit('free')` should be done.
	request.destroy();
});
```

Reply:
Ah, you are correct! Thanks

Reply:
Actually I'm unsure how we could detect whether a socket can or cannot be re-used. I don't think checking `headersSent` is enough.

This PR **does** fix it, at least in the case where `socket` has not yet been assigned to the request object.

Reply:
Updated PR with a TODO for future PRs.

Reply:
> Actually I'm unsure how we could detect whether a socket can or cannot be re-used. I don't think checking headersSent is enough.

It is. `request.flushHeaders()`, `request.write(...)` and `request.end()` both call `request._implicitHeader()` which sets `request._header` (and sends it). `request.headersSent` points to `!!request._header`.

Reply:
What about `_writeRaw`?

Reply:
It is called by `_send` which is called by either `_storeHeader` (which is called by `_implicitHeader`), `_write` (which is called by `write`) or `flushHeaders`.

`_storeHeader` is called by the `ClientRequest` instance when the `options.headers` is an array of there is a `expect` header defined.

https://github.com/nodejs/node/blob/5688eb8f83646a8c4896d5b81086377eb5092b3e/lib/_http_client.js#L275-L282

> It is called by `_send` 

This function in particular sets the `_headersSent` property.

Reply:
Ok, updated per your suggestion.